src/base/Pool.hh

```diff
@@ -0,0 +1,219 @@
+// Copyright 2021 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file Pool.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "NumericLimits.hh"
+#include "Types.hh"
+
+// Proxy for defining specializations in a separate header that device-only
+// code can omit
+#define POOL_HOST_HEADER 1
+#if POOL_HOST_HEADER
+#    include <vector>
+#    include "DeviceVector.hh"
+#endif
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+//! Data ownership flag
+enum class Ownership
+{
+    value,           //!< Ownership of the data, only on host
+    reference,       //!< Mutable reference to the data
+    const_reference, //!< Immutable reference to the data
+};
+
+//! Size/offset type for pool
+using pool_size_type = unsigned int;
+
+CELER_CONSTEXPR_FUNCTION size_type max_pool_size()
+{
+    return static_cast<size_type>(numeric_limits<pool_size_type>::max());
+}
+
+namespace detail
+{
+template<typename T, Ownership W>
+struct ownership_traits
+{
+    using SpanT          = Span<T>;
+    using reference_type = T&;
+};
+
+template<typename T>
+struct ownership_traits<T, Ownership::const_reference>
+{
+    using SpanT          = Span<const T>;
+    using reference_type = const T&;
+};
+} // namespace detail
+
+//---------------------------------------------------------------------------//
+template<class T>
+class PoolRange
+{
+  public:
+    PoolRange(pool_size_type start, pool_size_type stop)
+        : start_(start), stop_(stop)
+    {
+        CELER_EXPECT(start_ <= stop_);
+    }
+
+    //!@{
+    //! Range of indices in the corresponding pool
+    CELER_CONSTEXPR_FUNCTION pool_size_type start() const { return start_; }
+    CELER_CONSTEXPR_FUNCTION pool_size_type stop() const { return stop_; }
+    //!@}
+
+    //! Number of elements
+    CELER_CONSTEXPR_FUNCTION pool_size_type size() const
+    {
+        return start_ - stop_;
+    }
+
+  private:
+    pool_size_type start_{};
+    pool_size_type stop_{};
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Storage and access to subspans of a contiguous array.
+ *
+ * Pools are constructed incrementally on the host, then copied (along with
+ * their associated PoolRanges) to device.
+ */
+template<class T, Ownership W, MemSpace M>
+class Pool
+{
+  public:
+    using SpanT = typename detail::ownership_traits<T, W>::SpanT;
+    using reference_type =
+        typename detail::ownership_traits<T, W>::reference_type;
+
+    Pool& operator=(const Pool& other) = default;
+
+    //! Assign from another pool in the same memory space
+    template<Ownership W2>
+    Pool& operator=(const Pool<T, W2, M>& other)
```

Comment:
Unless I missed specialization for 'const reference' it seems the following is (unfortunately) legal:
```
Pool<sometype, const_reference, something> good;
Pool<sometype, reference, something> bad = good; // oups :)
```

Reply:
Not directly, but I’m pretty sure the pointer conversion from const T to T will fail in the depths of Span. That’s why I had to add the extra const reference option in the first place, I’m afraid

Reply:
Make sense.  Then it becomes the same challenge as https://github.com/celeritas-project/celeritas/pull/131/files/97c1edd28fa6b3cb5febb545d828d6555f929ff4#r568671286